### PR TITLE
Add cli option to group reporter output. Closes #38

### LINKED ISSUE #38: Add CLI option to group reporter output
## Description

The current output is displayed by the reporter as individual files without any grouping. Some users may want to have a customized grouping of the files in the output - for example grouping by file type, directory, pass/failed, etc. The options should be able to be combined/nested such as grouping by filetype within a directory

### PR DESCRIPTION
**Pull request by jcsmurph**
Added the option to group the report output by directory, filetype, or pass/fail using `-groupby`. The option can use one or more commands to organize the report output.

### CODE REVIEW

**Review by kehoecj — COMMENTED** [review]
@jcsmurph Thank you for the PR! I ran the Github action and it looks like there are a couple `go fmt` issues

File: README.md [hunk 1]
```diff
@@ -46,7 +46,7 @@ There are several ways to install the config file validator tool
 
 We offer alpine, ubuntu, and scratch containers
 
-#### Apline 
+#### Apline
```
Comment (jd4235) [inline README.md:49]:
```suggestion
#### Alpine
```

File: pkg/cli/cli_test.go [hunk 2]
```diff
@@ -10,6 +10,7 @@ import (
 func Test_CLI(t *testing.T) {
 	searchPath := "../../test"
 	excludeDirs := []string{"subdir", "subdir2"}
+    groupBy := []string{}
```
Comment (jd4235) [inline pkg/cli/cli_test.go:13]: Indent appears off. Align with previous line.

File: pkg/cli/cli_test.go [hunk 3]
```diff
@@ -19,6 +20,7 @@ func Test_CLI(t *testing.T) {
 	cli := Init(
 		WithFinder(fsFinder),
 		WithReporter(stdoutReporter),
+        WithGroupOutput(groupBy),
```
Comment (jd4235) [inline pkg/cli/cli_test.go:23]: Indent appears off. Align with previous line.

File: pkg/cli/group_output.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+	return reportByDirectory
+}
+
+func GroupBy(reports []reporter.Report, groupBy []string) []reporter.Report {
+    // Iterate through groupBy in reverse order
+    // This will make the first command the primary grouping
+    for i := len(groupBy)-1; i >= 0; i--  {
+		switch groupBy[i] {
+		case "pass/fail":
```
Comment (jd4235) [inline pkg/cli/group_output.go:85]: using a forward slash in a CLI argument may present problems when passing args into CLI from yml or other sources. Suggest changing the slash to a dash throughout.

**Comment by jcsmurph** [comment]
@kehoecj Seems my indentation is off in a few places. I will fix those along with @jd4235 suggestions once the review is completed.

**Comment by jd4235** [comment]
I'll re-review once comments are incorporated

**Comment by jcsmurph** [comment]
@kehoecj Hopefully it is good to go now. Seems I need to make some changes to my workspace config for Go formatting 😅

File: pkg/cli/group_output.go (hunk 4 above)
Comment (jcsmurph) [reply pkg/cli/group_output.go]: Changed pass/fail to pass-fail

File: cmd/validator/validator.go [hunk 5]
```diff
@@ -71,6 +73,7 @@ func getFlags() (validatorConfig, error) {
 	excludeFileTypesPtr := flag.String("exclude-file-types", "", "A comma separated list of file types to ignore")
 	depthPtr := flag.Int("depth", 0, "Depth of recursion for the provided search paths. Set depth to 0 to disable recursive path traversal")
 	versionPtr := flag.Bool("version", false, "Version prints the release version of validator")
+	groupOutputPtr := flag.String("groupby", "", "Group output by file type, directory, pass/fail")
```
Comment (jd4235) [inline cmd/validator/validator.go:76]:
```suggestion
	groupOutputPtr := flag.String("groupby", "", "Group output by file type, directory, pass-fail")
```

File: pkg/cli/cli_test.go [hunk 6]
```diff
@@ -70,3 +72,29 @@ func Test_CLIBadPath(t *testing.T) {
 		t.Errorf("Exit status was not 1")
 	}
 }
+
+func Test_CLIWithGroup(t *testing.T) {
+    searchPath := "../../test"
+    excludeDirs := []string{"subdir", "subdir2"}
+    groupOutput := []string{"pass/fail", "directory", "filetype"}
```
Comment (jd4235) [inline pkg/cli/cli_test.go:79]:
```suggestion
    groupOutput := []string{"pass-fail", "directory", "filetype"}
```

File: cmd/validator/validator.go (hunk 5 above)
Comment (jcsmurph) [reply cmd/validator/validator.go]: Done

File: pkg/cli/cli_test.go (hunk 6 above)
Comment (jcsmurph) [reply pkg/cli/cli_test.go]: Done

**Review by kehoecj — CHANGES_REQUESTED** [review]
I haven't gone through the code yet but I did some local testing of the changes and have some findings and recommended changes:

### Issues

There is an issue when all of the tests are in the same parent directory and `--groupby=directory` is used:

```
.\validator.exe --groupby=directory .\test\fixtures\
panic: runtime error: index out of range [-1]

goroutine 1 [running]:
github.com/Boeing/config-file-validator/pkg/cli.GroupByDirectory({0xc00025a000, 0x18, 0xd341a?})
        C:/src/github.com/pr/grouping/config-file-validator/pkg/cli/group_output.go:65 +0x478
github.com/Boeing/config-file-validator/pkg/cli.GroupBy({0xc00025a000?, 0xc000268000?, 0xc00011bc50?}, {0xc0000223d0, 0x35a700?, 0xc000008720?})
        C:/src/github.com/pr/grouping/config-file-validator/pkg/cli/group_output.go:90 +0xb7
github.com/Boeing/config-file-validator/pkg/cli.CLI.Run({{0x409780?, 0xc000164e00?}, {0x4097c0?, 0x63ea20?}})
        C:/src/github.com/pr/grouping/config-file-validator/pkg/cli/cli.go:101 +0x38a
main.mainInit()
        C:/src/github.com/pr/grouping/config-file-validator/cmd/validator/validator.go:203 +0x56c
main.main()
        C:/src/github.com/pr/grouping/config-file-validator/cmd/validator/validator.go:212 +0x13
 ```

#### Recommended Changes
1.  Grouping by filetype does not group yml and yaml together (I could be talked out of this being required though) and should probably be sorted alphabetically
2. The output isn't being grouped by the designators but rather ordered by. For example if you passed `--groupby=pass-fail,filetype` I think something like this would make the output more readable:

```
Passed
  JSON
     ✓ test\fixtures\exclude-file-types\excluded.json
  HCL
    ✓ test\fixtures\good.hcl
Failed
  JSON
    × test\fixtures\subdir\bad.json
        error: Error at line 3 column 14: invalid character ':' after array element
```
and the JSON output would also reflect that structure:

```
{
  "files": {
    "passed": {
      "json": [
        {
          "path": "test/fixtures/good.toml",
          "status": "passed"
        }
      ]
    },
    "failed": {
      "hcl": [
        {
          "path": "test/fixtures/subdir2/bad.hcl",
          "status": "failed",
          "error": "error at line 1 column 1: :1,1-2: Invalid argument name; Argument names must not be quoted."
        }
      ]
    }
  }
}
```

**Comment by jcsmurph** [comment]
@kehoecj The changes requested are going to take me a bit. It may be worthwhile to close out this PR and I can resubmit for review once I am done. Thoughts?

**Comment by kehoecj** [comment]
> @kehoecj The changes requested are going to take me a bit. It may be worthwhile to close out this PR and I can resubmit for review once I am done. Thoughts?

Whatever is easier for you! I'm happy to leave it open while you work through the changes but if you'd rather close it and open a new one when finished that works too.

**Comment by jcsmurph** [comment]
@kehoecj We can leave it open. It should be ready for another review now.

I wasn't able to duplicate the index error on my end so let me know if you run into again. If so, I will continue testing.

File: pkg/reporter/json_reporter.go [hunk 7]
```diff
@@ -67,3 +88,206 @@ func (jr JsonReporter) Print(reports []Report) error {
...
+	totalFailed := 0
+	report.Files = make(map[string][]fileStatus)
+	report.Summary = make(map[string][]summary)
+
+	for group, reports := range groupReports {
+		report.Files[group] = make([]fileStatus, 0)
+		report.Summary[group] = make([]summary, 0)
+		currentPassed = 0
+		currentFailed = 0
+		for _, r := range reports {
```
Comment (jd4235) [inline pkg/reporter/json_reporter.go:111]: This for loop and the next appear to be duplicated in this and the other new print functions. I would prefer to eliminate duplicate code. Please refactor the duplicate code into a separate utility function.

File: pkg/reporter/stdout_reporter.go [hunk 8]
```diff
@@ -31,6 +31,113 @@ func (sr StdoutReporter) Print(reports []Report) error {
...
+			fmt.Printf("    Summary: %d succeeded, %d failed\n\n", successCount, failureCount)
+		}
+	}
+
+	fmt.Printf("Total Summary: %d succeeded, %d failed\n", totalSuccessCount, totalFailureCount)
+
+	return nil
+}
+
+func (sr StdoutReporter) PrintTripleGroup(groupReport map[string]map[string]map[string][]Report, groupOutput []string) error {
```
Comment (jackswiney) [inline pkg/reporter/stdout_reporter.go:103]: Same as above, comment header needed.

**Comment by jcsmurph** [comment]
@jd4235 I created the utility function and cleaned up the JSON reporter. Let me know if that is closer to what is wanted.

**Comment by kehoecj** [comment]
@jcsmurph Looks like we have some merge conflicts - can you take care of them and we'll do a final review. We'd like to get all your good work merged in

**Comment by jcsmurph** [comment]
Should be good to go now. It was a quick change to the standard out testing file.

**Comment by jcsmurph** [comment]
Looks like the new JUnit reporter needs to implement the groupBy print functions (either working or returning nil) or I could remove the groupBy print functions from the interface.

@kehoecj What do you think?

**Comment by kehoecj** [comment]
> Looks like the new JUnit reporter needs to implement the groupBy print functions (either working or returning nil) or I could remove the groupBy print functions from the interface.
> 
> @kehoecj What do you think?

@jcsmurph Good question - I think Junit should be excluded from the groupby functionality. I don't see a lot of benefit to adding it since JUnit is really for processing. JUnit also has a pretty strict structure so I'm not even sure how feasible grouping it would be @jd4235 WDTY?

**Comment by jd4235** [comment]
> > Looks like the new JUnit reporter needs to implement the groupBy print functions (either working or returning nil) or I could remove the groupBy print functions from the interface.
> > @kehoecj What do you think?
> 
> @jcsmurph Good question - I think Junit should be excluded from the groupby functionality. I don't see a lot of benefit to adding it since JUnit is really for processing. JUnit also has a pretty strict structure so I'm not even sure how feasible grouping it would be @jd4235 WDTY?

I agree that JUnit should be excluded from the grouping.

**Comment by jcsmurph** [comment]
> I agree that JUnit should be excluded from the grouping.

Sounds good. The options I am seeing are:

1. Remove the groupBy functions from the interface. Refactor the functions.
2. Implement the groupBy functions for JUnit but return nil/an error saying it is not supported. 
3. A third option I have not considered.

I should probably add a validation check for the flags if a user passes JUnit for the reporter, they cannot add a groupBy. That way the user gets the error immediately rather than after the validator has ran.

**Comment by kehoecj** [comment]
> > I agree that JUnit should be excluded from the grouping.
> 
> Sounds good. The options I am seeing are:
> 
> 1. Remove the groupBy functions from the interface. Refactor the functions.
> 2. Implement the groupBy functions for JUnit but return nil/an error saying it is not supported.
> 3. A third option I have not considered.
> 
> I should probably add a validation check for the flags if a user passes JUnit for the reporter, they cannot add a groupBy. That way the user gets the error immediately rather than after the validator has ran.

Option 1 seems the cleanest to me. I agree with the validation check if a user passes groupby and junit

**Comment by jcsmurph** [comment]
Changes have been made and testing looks good on my end.

I am noticing now with the support for case mixed filetypes, they get sorted differently depending on the case. E.g. Csv is sorted differently than csv, CsV, etc. Did you want me to take a look into changing that as part of this PR?

**Comment by jcsmurph** [comment]
I really need to figure out what is wrong with my local config and gofmt...

Sorry, you can try the workflow again. It should validate this time.

**Comment by kehoecj** [comment]
@jcsmurph I did a final functional test today running through several scenarios. There is currently an issue with `--groupby directory` functionality. For example running `.\validator.exe -groupby directory .\test` displays this output:

```
PS C:\Users\se456c\src\github.com\pr\grouping\config-file-validator> .\validator.exe -groupby directory .\test
/
    ✓ test\fixtures\exclude-file-types\excluded.json
    ✓ test\fixtures\exclude-file-types\not-excluded.toml
    ✓ test\fixtures\good.csv
    ✓ test\fixtures\good.hcl
```

This was definitely working before so it may be something that was introduced with one of the latest sets of changes. All my other functional tests passed so this should be the last issue before I'm ready to merge

**Comment by kehoecj** [comment]
> @kehoecj If I understand correctly, it is grouping the test\fixtures\ and test\fixtures\exclude-file-types\ directories together?

It's grouping everything under `/`. For example even running this commend `PS C:\Users\se456c\src\github.com\pr\grouping\config-file-validator> .\validator.exe -groupby directory .\test C:\ProgramData\
/` groups all the output from two totally separate places under `/`

**Comment by jcsmurph** [comment]
Got it. Seems to be Windows specific as I am not having the issue with my Linux setup but am with Windows. I will take a look.

**Comment by jcsmurph** [comment]
@kehoecj I made an update and it is working on my Windows machine now. Let me know if you see any other issues.

**Review by kehoecj — APPROVED** [review]

**Comment by kehoecj** [comment]
> @kehoecj I made an update and it is working on my Windows machine now. Let me know if you see any other issues.

LGTM! Thanks for the quick turnaround

**Comment by jcsmurph** [comment]
@kehoecj I added 2 tests for the directory GroupBy function to cover Windows file paths and others. That should hopefully get the test coverage back in the threshold.

**Comment by kehoecj** [comment]
> @kehoecj I added 2 tests for the directory GroupBy function to cover Windows file paths and others. That should hopefully get the test coverage back in the threshold.

That did it - thanks!

**Comment by kehoecj** [comment]
Merged - thank you for the great work on this PR @jcsmurph !

**Comment by jcsmurph** [comment]
Awesome! I will keep an eye out for other issues I may want to tackle.

**Pull request merged**